File: code_examples/DesignPatterns/state.kt
// Updated State interface
interface MarioState {
    fun stateBehaviour(): String
    fun accessorize(): String
}

class BaseMode : MarioState {
    override fun stateBehaviour(): String {
        return "It's-a me, Mario!"
    }
    override fun accessorize(): String {
        return "Blue Overalls"
    }
}

class DriverMode : MarioState {
    override fun stateBehaviour(): String {
        return "Ready to race!"
    }
    override fun accessorize(): String {
        return "Race Car"
    }
}

class DoctorMode : MarioState {
    override fun stateBehaviour(): String {
        return "Dr. Mario at your service!"
    }
    override fun accessorize(): String {
        return "Stethoscope"
    }
}

class TennisMode : MarioState {
    override fun stateBehaviour(): String {
        return "Ready to serve!"
    }
    override fun accessorize(): String {
        return "Tennis Racket"
    }
}

class FightMode : MarioState {
    override fun stateBehaviour(): String {
        return "Combat stance activated!"
    }
    override fun accessorize(): String {
        return "Fire Flower"
    }
}

class Mario(var current: MarioState) {
    fun text(): String {
        return "${current.stateBehaviour()} \n Now equipped with a ${current.accessorize()}"
    }
}

fun main() {
    val baseMario = Mario(BaseMode())
    val raceMario = Mario(DriverMode())
    val doctorMario = Mario(DoctorMode())
    val tennisMario = Mario(TennisMode())
    val fightMario = Mario(FightMode())

    println(baseMario.text())
    println(raceMario.text())
    println(doctorMario.text())
    println(tennisMario.text())
    println(fightMario.text())
}

File: code_examples/DesignPatterns/State.js
class MarioState {
    stateBehaviour() {}
    accessorize() {}
}

class BaseMode extends MarioState {
    stateBehaviour() {
        return "It's-a me, Mario!";
    }

    accessorize() {
        return "Blue Overalls";
    }
}

class DriverMode extends MarioState {
    stateBehaviour() {
        return "Ready to race!";
    }

    accessorize() {
        return "Race Car";
    }
}

class DoctorMode extends MarioState {
    stateBehaviour() {
        return "Dr. Mario at your service!";
    }

    accessorize() {
        return "Stethoscope";
    }
}

class TennisMode extends MarioState {
    stateBehaviour() {
        return "Ready to serve!";
    }

    accessorize() {
        return "Tennis Racket";
    }
}

class FightMode extends MarioState {
    stateBehaviour() {
        return "Combat stance activated!";
    }

    accessorize() {
        return "Fire Flower";
    }
}

class Mario {
    constructor(current) {
        this.current = current;
    }

    text() {
        return `${this.current.stateBehaviour()} \n Now equipped with a ${this.current.accessorize()}`;
    }
}

const baseMario = new Mario(new BaseMode());
const raceMario = new Mario(new DriverMode());
const doctorMario = new Mario(new DoctorMode());
const tennisMario = new Mario(new TennisMode());
const fightMario = new Mario(new FightMode());

console.log(baseMario.text());
console.log(raceMario.text());
console.log(doctorMario.text());
console.log(tennisMario.text());
console.log(fightMario.text());
